# 5장 연습문제
---------

# Q1
-----
다음은 Calculater 클래스이다.

In [1]:
class Calculator:
    def __init__(self):
        self.value = 0

    def add(self, val):
        self.value += val

위 클래스를 상속하는 UpgradeCalculator를 만들고 값을 뺄 수 있는 minus 메서드를 추가해 보자. 즉 다음과 같이 동작하는 클래스를 만들어야 한다.

cal = UpgradeCalculator()
cal.add(10)
cal.minus(7)

print(cal.value) # 10에서 7을 뺀 3을 출력

In [3]:
class UpgradeCalculator(Calculator):
    def minus(self,val):
        self.value -= val

In [5]:
cal = UpgradeCalculator()
cal.add(10)
cal.minus(7)

print(cal.value)

3


# Q2
------
객체변수 value가 100 이상의 값은 가질 수 없도록 제한하는 MaxLimitCalculator 클래스를 만들어 보자. 즉 다음과 같이 동작해야 한다.

cal = MaxLimitCalculator()
cal.add(50) # 50 더하기
cal.add(60) # 60 더하기

print(cal.value) # 100 출력

단 반드시 다음과 같은 Calculator 클래스를 상속해서 만들어야 한다.

In [7]:
class Calculator:
    def __init__(self):
        self.value = 0

    def add(self, val):
        self.value += val

In [8]:
class MaxLimitCalculator(Calculator):
    def add(self,val):
        self.value += val
        if self.value>100:
            self.value=100

In [9]:
cal = MaxLimitCalculator()
cal.add(50)
cal.add(60)

print(cal.value)

100


# Q3
-----
다음 결과를 예측해 보자.  

하나,

all([1, 2, abs(-3)-3])

둘.

chr(ord('a')) == 'a'

abs()는 절댓값으로 바꿔주는 함수이므로 all([1,2,0])이 되고 0은 거짓으로 인식되기 떄문에 all()함수에 의해서 False가 나온다.

In [10]:
all([1, 2, abs(-3)-3])

False

ord()는 문자의 아스키 코드 값을 돌려주는 함수이고, chr()은 아스키 코드 값을 입력받아서 그 코드에 해당하는 문자를 출력하는 함수이므로 chr(ord())은 입력한 값을 그대로 돌려주는 것을 의미한다. 따라서 입력을 'a'를 하였고 그것이 'a'와 같은지를 판단하는 것이기에 True가 나온다.

In [11]:
chr(ord('a')) == 'a'

True

# Q4
-----
filter와 lambda를 사용하여 리스트 [1,-2,3,-5,8,-3]에서 음수를 모두 제거해 보자.

In [12]:
list(filter(lambda x:x>0, [1,-2,3,-5,8,-3]))

[1, 3, 8]

# Q5
-----
234라는 10진수의 16진수는 다음과 같이 구할 수 있다.

In [13]:
hex(234)

'0xea'

이번에는 반대로 16진수 문자열 0xea를 10진수로 변경해 보자.

In [14]:
int('0xea',16)

234

# Q6
-----
map와 lambda를 사용하여 [1,2,3,4] 리스트의 각 요솟값에 3이 곱해진 리스트 [3,6,9,12]를 만들어 보자.

In [17]:
list(map(lambda x:x*3, [1,2,3,4]))

[3, 6, 9, 12]

# Q7
-----
다음 리스트의 최댓값과 최솟값의 합을 구해보자.

[-8, 2, 7, 5, -3, 5, 0, 1]

In [20]:
a=[-8, 2, 7, 5, -3, 5, 0, 1]
max(a)+min(a)

-1

# Q8
-----
17/3의 결과는 다음과 같다.

In [21]:
17 / 3

5.666666666666667

위와 같은 결괏값 5.666666666666667을 소숫점 4자리까지만 반올림하여 표시해 보자.

In [22]:
round(17/3,4)

5.6667

# Q9
------
다음과 같이 실행할 때 입력값을 모두 더하여 출력하는 스크립트(C:\doit\myargv.py)를 작성해 보자.

C:\> cd doit
C:\doit> python myargv.py 1 2 3 4 5 6 7 8 9 10
55

import sys

numbers = sys.argv[1:]

result = 0
for number in numbers:
    result += int(number)
print(result)

# Q10
-----
os 모듈을 사용하여 다음과 같이 동작하도록 코드를 작성해 보자.

1. `C:\doit` 디렉터리로 이동한다.
2. dir 명령을 실행하고 그 결과를 변수에 담는다.
3. dir 명령의 결과를 출력한다.

In [33]:
import os
os.chdir("c:/doit")
result = os.popen("dir")
print(result.read())

 C 드라이브의 볼륨에는 이름이 없습니다.
 볼륨 일련 번호: 2E8A-F43B

 c:\doit 디렉터리

2019-10-10  오전 03:33    <DIR>          .
2019-10-10  오전 03:33    <DIR>          ..
               0개 파일                   0 바이트
               2개 디렉터리  46,986,276,864 바이트 남음



# Q11
-----
glob 모듈을 사용하여 `C:\doit` 디렉터리의 파일 중 확장자가 .py인 파일만 출력하는 프로그램을 작성해 보자.

In [34]:
import glob
glob.glob("c:/doit/*.py")

[]

# Q12
-----
time 모듈을 사용하여 현재 날짜와 시간을 다음과 같은 형식으로 출력해 보자.

2018/04/03 17:20:32

In [35]:
import time
time.strftime("%Y/%m/%d %H:%M:%S")

'2019/10/10 03:42:32'

# Q13
-----
random 모듈을 사용하여 로또 번호(1~45 사이의 숫자 6개)를 생성해 보자(단 중복된 숫자가 있으면 안 됨).

In [36]:
import random

result = []
while len(result) < 6:
    num = random.randint(1, 45)
    if num not in result:
        result.append(num)

print(result)

[6, 17, 33, 3, 12, 10]
